### PR TITLE
[BEAM-1623] Transform Reshuffle directly in Spark runner

### PR DESCRIPTION
**Pull request by aviemzur**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by aviemzur** [comment]
R: @amitsela

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/GroupCombineFunctions.java [hunk 1]
```diff
@@ -203,4 +203,26 @@
...
+   * {@link org.apache.beam.sdk.util.Reshuffle} for the Spark runner.
+   */
+  public static <K, V> JavaRDD<WindowedValue<KV<K, V>>> reshuffle(
+      JavaRDD<WindowedValue<KV<K, V>>> rdd,
+      Coder<K> keyCoder,
+      WindowedValueCoder<V> wvCoder) {
+
+    // Use coders to convert objects in the PCollection to byte arrays, so they
+    // can be transferred over the network for the shuffle.
+    return rdd
```
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/GroupCombineFunctions.java]: I'm assuming you're "Reifying" to preserve the window before transforming to a pair ? I know that this is how Reshuffle is implemented in Beam, but I wonder if the Spark implementation really cares about it ?
Isn't it enough to perform: 
`JavaRDD<WindowedValue<T>>` -> `JavaRDD<byte[]>` (using a `FullWindowedValueCoder`) -> repartition -> and back to deserialized representation ?

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TranslationUtils.java [hunk 2]
```diff
@@ -159,6 +159,18 @@ public T2 call(Tuple2<T1, T2> v1) throws Exception {
...
       }
 
+  /** Extract window from a {@link KV} with {@link WindowedValue} value. */
+  static <K, V> Function<KV<K, WindowedValue<V>>, WindowedValue<KV<K, V>>> toKVByWindowInValue() {
+    return new Function<KV<K, WindowedValue<V>>, WindowedValue<KV<K, V>>>() {
+      @Override public WindowedValue<KV<K, V>> call(KV<K, WindowedValue<V>> kwv) throws Exception {
+        K k = kwv.getKey();
+        WindowedValue<V> wv = kwv.getValue();
+        V v = wv.getValue();
+        return WindowedValue.of(KV.of(k, v), wv.getTimestamp(), wv.getWindows(), wv.getPane());
```
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TranslationUtils.java]: Try: ` wv.withValue(KV.of(kv.getKey(), wv.getValue()))` instead.
This is used in a couple of other places such as [combinePerKey](https://github.com/apache/beam/blob/master/runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java#L327) so you could look them up and use this utility there as well.

**Comment by amitsela** [comment]
Reshuffle looks good, just a minor comment on the util you added since I think you can use it to squash a few more lines of code in the runner.

**Comment by aviemzur** [comment]
@amitsela addressed your comment in my latest commit.

**Comment by amitsela** [comment]
Run Spark RunnableOnService

**Comment by aviemzur** [comment]
Retest this please.

**Comment by amitsela** [comment]
LGTM

**Pull request closed without merging**